controllers/topology/cluster_controller_test.go

```diff
@@ -0,0 +1,377 @@
+	g.Eventually(func() bool {
+		return assertClusterReconcile(ns.Name, clusterName, actualCluster)
+	}, timeout).Should(BeTrue())
+
+	// Check if InfrastructureCluster has been created and has the correct labels and annotations.
+	g.Eventually(func() bool {
+		return assertInfrastructureClusterReconcile(actualCluster)
+	}, timeout).Should(BeTrue())
+
+	// Check if ControlPlane has been created and has the correct version, replicas, labels and annotations.
+	g.Eventually(func() bool {
+		return assertControlPlaneReconcile(actualCluster, clusterClass)
+	}, timeout).Should(BeTrue())
+	g.Eventually(func() bool {
+		return assertMachineDeploymentsReconcile(actualCluster)
+	}, timeout).Should(BeTrue())
+}
+
+// assertClusterReconcile checks if the Cluster object:
+// 1) Has its InfrastructureReference and ControlPlane reference added correctly.
+// 2) InfrastructureReference and ControlPlaneRef have the expected Group, Version and Kind.
+func assertClusterReconcile(namespace, clusterName string, cluster *clusterv1.Cluster) bool {
+	key := client.ObjectKey{Name: clusterName, Namespace: namespace}
+	if err := env.Get(ctx, key, cluster); err != nil {
+		return false
+	}
+	// Check if relevant managed topology labels are present.
+	if ok := assertClusterTopologyOwnedLabel(cluster); !ok {
+		return false
+	}
+	// Check if InfrastructureRef exists and is of the expected Kind and APIVersion.
+	if ok := referenceExistsWithCorrectKindAndAPIVersion(cluster.Spec.InfrastructureRef,
+		builder.GenericInfrastructureClusterKind,
+		builder.InfrastructureGroupVersion); !ok {
+		log.Infof("We failed on infraref %v", cluster)
+		return false
+	}
+
+	// Check if ControlPlaneRef exists is of the expected Kind and APIVersion.
+	if ok := referenceExistsWithCorrectKindAndAPIVersion(cluster.Spec.ControlPlaneRef,
+		builder.GenericControlPlaneKind,
+		builder.ControlPlaneGroupVersion); !ok {
+		log.Infof("We failed on controlPlane ref %v", cluster)
+		return false
+	}
+	log.Infof("Cluster passed with %v", cluster)
+	return true
+}
+
+// assertInfrastructureClusterReconcile checks if the infrastructureCluster object:
+// 1) Is created correctly.
+// 2) Has the correct labels and annotations.
+func assertInfrastructureClusterReconcile(cluster *clusterv1.Cluster) bool {
+	_, ok := getAndAssertLabelsAndAnnotations(*cluster.Spec.InfrastructureRef, cluster.Name)
+	return ok
+}
+
+// assertControlPlaneReconcile checks if the ControlPlane object:
+// 1) Is created correctly.
+// 2) Has the correct labels and annotations.
+// 3) If it requires ControlPlane Infrastructure and if so:
+//		i) That the infrastructureMachineTemplate is created correctly.
+//      ii) That the infrastructureMachineTemplate has the correct labels and annotations
+func assertControlPlaneReconcile(cluster *clusterv1.Cluster, clusterClass *clusterv1.ClusterClass) bool {
+	cp, ok := getAndAssertLabelsAndAnnotations(*cluster.Spec.ControlPlaneRef, cluster.Name)
```

Comment:
A few empty lines to structure this func would be good (same in other funcs below)